File: app/src/main/java/com/example/mealrecipeapp/data/repository/AppRepository.java
package com.example.mealrecipeapp.data.repository;

import android.content.SharedPreferences;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.mealrecipeapp.data.local.database.AppDatabase;
import com.example.mealrecipeapp.data.local.database.RecipeDao;
import com.example.mealrecipeapp.data.local.entity.RecipeEntity;
import com.example.mealrecipeapp.data.remote.network.ApiService;
import com.example.mealrecipeapp.data.remote.response.AddMealPlanResponse;
import com.example.mealrecipeapp.data.remote.response.ConnectUserResponse;
import com.example.mealrecipeapp.data.remote.response.GetMealPlanResponse;
import com.example.mealrecipeapp.data.remote.response.GetRecipeResponse;
import com.example.mealrecipeapp.data.remote.response.MealPlan;
import com.example.mealrecipeapp.data.remote.response.MealPlanValue;
import com.example.mealrecipeapp.data.remote.response.Recipe;
import com.example.mealrecipeapp.data.remote.response.RecipeInformation;
import com.example.mealrecipeapp.utils.Constants;
import com.example.mealrecipeapp.utils.Resource;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class AppRepository {
    private final ApiService apiService;
    private final SharedPreferences sharedPreferences;
    private final RecipeDao recipeDao;
    private final FirebaseFirestore firestoreDB = FirebaseFirestore.getInstance();
    public AppRepository(ApiService apiService, SharedPreferences sharedPreferences, RecipeDao recipeDao) {
        this.apiService = apiService;
        this.sharedPreferences = sharedPreferences;
        this.recipeDao = recipeDao;
    }

    public LiveData<Resource<String>> saveUser(String email, String name, String image) {
        final MutableLiveData<Resource<String>> resource = new MutableLiveData<>();
        resource.postValue(Resource.loading(null));

        SharedPreferences.Editor editor = sharedPreferences.edit();

        DocumentReference docRef = firestoreDB.collection("users").document(email);
        // check data in firestore
        docRef.get().addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                DocumentSnapshot document = task.getResult();
                if (document.exists()) {
                    // data exist in firestore
                    // save data user to shared preferences
                    editor.putString("email", email);
                    editor.putString("name", name);
                    editor.putString("image", image);
                    editor.putString("username", document.getString("username"));
                    editor.putString("hash", document.getString("hash"));
                    editor.apply();
                    resource.postValue(Resource.success("Success Sign In"));
                } else {
                    // data not exist in firestore
                    // create data user from api
                    Call<ConnectUserResponse> call = apiService.connectUser(Constants.API_KEY);
                    call.enqueue(new Callback<ConnectUserResponse>() {
                        @Override
                        public void onResponse(@NonNull Call<ConnectUserResponse> call, @NonNull Response<ConnectUserResponse> response) {
                            if (response.body() != null && response.isSuccessful()) {
                                // success create data user from api
                                // save data user to firestore
                                Map<String, Object> user = new HashMap<>();
                                user.put("username", response.body().getUsername());
                                user.put("hash", response.body().getHash());

                                docRef.set(user)
                                    .addOnSuccessListener(aVoid -> {
                                        // success save data user to firestore
                                        // save data user to shared preferences
                                        editor.putString("email", email);
                                        editor.putString("name", name);
                                        editor.putString("image", image);
                                        editor.putString("username", response.body().getUsername());
                                        editor.putString("hash", response.body().getHash());
                                        editor.apply();
                                        resource.postValue(Resource.success("Success Sign In"));
                                    })
                                    .addOnFailureListener(e ->  resource.postValue(Resource.error("Fail Sign In", null)));
                            }
                        }

                        @Override
                        public void onFailure(@NonNull Call<ConnectUserResponse> call, @NonNull Throwable t) {
                            resource.postValue(Resource.error("Fail Sign In", null));
                        }
                    });
                }
            } else {
                resource.postValue(Resource.error("Fail Sign In", null));
            }
        });

        return resource;
    }

    public void deleteUser() {
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.remove("email");
        editor.remove("name");
        editor.apply();

        AppDatabase.databaseWriteExecutor.execute(recipeDao::deleteAll);
    }

    public String getUserEmail() {
        return sharedPreferences.getString("email", "Email");
    }

    public String getUserName() {
        return sharedPreferences.getString("name", "Name");
    }
    public String getUserImage() {
        return sharedPreferences.getString("image", "");
    }

    public LiveData<Resource<List<Recipe>>> getRecipes(String query) {
        final MutableLiveData<Resource<List<Recipe>>> resource = new MutableLiveData<>();

        resource.postValue(Resource.loading(null));

        Call<GetRecipeResponse> call = apiService.getRecipes(Constants.API_KEY, 25, true, "popularity", query, true, true);

        call.enqueue(new Callback<GetRecipeResponse>() {
            @Override
            public void onResponse(@NonNull Call<GetRecipeResponse> call, @NonNull Response<GetRecipeResponse> response) {
                if (response.code() == 200 && response.body() != null) {
                    List<Recipe> recipeList = new ArrayList<>(response.body().getRecipes());
                    recipeList.removeIf(recipe -> recipe.getAnalyzedInstructions().isEmpty());
                    recipeList.removeIf(recipe -> recipe.getExtendedIngredients().isEmpty());

                    resource.postValue(Resource.success(recipeList));
                } else {
                    resource.postValue(Resource.error("Data Empty", null));
                }
            }

            @Override
            public void onFailure(@NonNull Call<GetRecipeResponse> call, @NonNull Throwable t) {
                resource.postValue(Resource.error(t.getMessage(), null));
            }
        });

        return resource;
    }

    public LiveData<Resource<String>> addMealPlan(Long date, int slot, Recipe recipe) {
        final MutableLiveData<Resource<String>> result = new MutableLiveData<>();
        result.postValue(Resource.loading(null));

        String username = sharedPreferences.getString("username", "");
        String hash = sharedPreferences.getString("hash", "");

        MealPlanValue valueBody = new MealPlanValue(recipe.getID(), recipe.getServings(), recipe.getTitle(), recipe.getImage(), recipe.getImageType());
        MealPlan body = new MealPlan(0L, (date/1000)+25200, slot, 0, "RECIPE", valueBody);

        Call<AddMealPlanResponse> call = apiService.addMealPlan(username, hash, Constants.API_KEY, body);

        call.enqueue(new Callback<AddMealPlanResponse>() {
            @Override
            public void onResponse(@NonNull Call<AddMealPlanResponse> call, @NonNull Response<AddMealPlanResponse> response) {
                result.postValue(Resource.success("Success Add Meal Plan"));
            }

            @Override
            public void onFailure(@NonNull Call<AddMealPlanResponse> call, @NonNull Throwable t) {
                result.postValue(Resource.error("Fail Add Meal Plan", null));
            }
        });

        return result;
    }

    public LiveData<Resource<RecipeInformation>> getRecipeInformation(Long id) {
        final MutableLiveData<Resource<RecipeInformation>> result = new MutableLiveData<>();

        result.postValue(Resource.loading(null));

        Call<RecipeInformation> call = apiService.getRecipeInformation(id, Constants.API_KEY);

        call.enqueue(new Callback<RecipeInformation>() {
            @Override
            public void onResponse(@NonNull Call<RecipeInformation> call, @NonNull Response<RecipeInformation> response) {
                if (response.code() == 200 & response.body() != null) {
                    AppDatabase.databaseWriteExecutor.execute(() -> {
                        RecipeInformation recipeInformation = response.body();
                        int favoriteCount = recipeDao.getFavoriteRecipeById(recipeInformation.getID());
                        if (favoriteCount > 0) {
                            recipeInformation.setFavorite(true);
                        }
                        result.postValue(Resource.success(recipeInformation));
                    });
                } else {
                    result.postValue(Resource.error("Fail Get Recipe Information", null));
                }
            }

            @Override
            public void onFailure(@NonNull Call<RecipeInformation> call, @NonNull Throwable t) {
                result.postValue(Resource.error(t.getMessage(), null));
            }
        });

        return result;
    }

    public LiveData<String> deleteMealPlan(Long id) {
        final MutableLiveData<String> result = new MutableLiveData<>();

        String username = sharedPreferences.getString("username", "");
        String hash = sharedPreferences.getString("hash", "");

        Call<AddMealPlanResponse> call = apiService.deleteMealPlan(username, id, hash, Constants.API_KEY);

        call.enqueue(new Callback<AddMealPlanResponse>() {
            @Override
            public void onResponse(@NonNull Call<AddMealPlanResponse> call, @NonNull Response<AddMealPlanResponse> response) {
                result.postValue("Success Delete Meal Plan");
            }

            @Override
            public void onFailure(@NonNull Call<AddMealPlanResponse> call, @NonNull Throwable t) {
                result.postValue("Fail Delete Meal Plan");
            }
        });

        return result;
    }

    public LiveData<Resource<List<MealPlan>>> getMealPlans(Date date) {
        final MutableLiveData<Resource<List<MealPlan>>> mealPlan = new MutableLiveData<>();

        mealPlan.postValue(Resource.loading(null));

        String username = sharedPreferences.getString("username", "");
        String hash = sharedPreferences.getString("hash", "");

        final String dateFormat  = "yyyy-MM-dd";
        final String formattedDate = new SimpleDateFormat(dateFormat, Locale.getDefault()).format(date);

        Call<GetMealPlanResponse> call = apiService.getMealPlans(username, formattedDate, hash, Constants.API_KEY);

        call.enqueue(new Callback<GetMealPlanResponse>() {
            @Override
            public void onResponse(@NonNull Call<GetMealPlanResponse> call, @NonNull Response<GetMealPlanResponse> response) {
                if (response.code() == 200 && response.body() != null) {
                    mealPlan.postValue(Resource.success(response.body().getItems()));
                } else {
                    mealPlan.postValue(Resource.error("No meals planned for that day", null));
                }
            }

            @Override
            public void onFailure(@NonNull Call<GetMealPlanResponse> call, @NonNull Throwable t) {
                mealPlan.postValue(Resource.error(t.getMessage(), null));
            }
        });

        return mealPlan;
    }

    public void addFavoriteRecipe(RecipeEntity recipeEntity) {
        AppDatabase.databaseWriteExecutor.execute(() -> recipeDao.insert(recipeEntity));
    }

    public void removeFavoriteRecipe(RecipeEntity recipeEntity) {
        AppDatabase.databaseWriteExecutor.execute(() -> recipeDao.delete(recipeEntity));
    }

    public LiveData<List<RecipeEntity>> getFavoriteRecipes() {
        return recipeDao.getFavoriteRecipes();
    }
}

File: app/src/main/java/com/example/mealrecipeapp/data/repository/AppRepository.kt
package com.example.mealrecipeapp.data.repository

import android.content.Context
import android.content.SharedPreferences
import android.provider.Settings
import android.view.accessibility.AccessibilityManager
import androidx.lifecycle.LiveData
import com.example.mealrecipeapp.data.local.database.RecipeDao
import com.example.mealrecipeapp.data.local.entity.RecipeEntity
import com.example.mealrecipeapp.data.remote.network.ApiService
import com.example.mealrecipeapp.data.remote.response.MealPlan
import com.example.mealrecipeapp.data.remote.response.MealPlanValue
import com.example.mealrecipeapp.data.remote.response.Recipe
import com.example.mealrecipeapp.data.remote.response.RecipeInformation
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.scottyab.rootbeer.RootBeer
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale


class AppRepository(
    private val context: Context,
    private val apiService: ApiService,
    private val sharedPreferences: SharedPreferences,
    private val recipeDao: RecipeDao,
    private val firestoreDB: FirebaseFirestore,
    private val crashlytics: FirebaseCrashlytics,
    private val auth: FirebaseAuth,
    private val rootBeer: RootBeer,
) {
    fun checkIsRooted(): Boolean {
        val isRooted =  rootBeer.isRooted
        if (isRooted) {
            crashlytics.recordException(Exception("Device Is Rooted"))
        }
        return isRooted
    }

    fun getCheckRootSetting(): Boolean {
        return sharedPreferences.getBoolean("checkRooted", true)
    }

    fun setCheckRootSetting(isChecked: Boolean) {
        val editor = sharedPreferences.edit()
        editor.putBoolean("checkRooted", isChecked)
        editor.apply()
    }

    private fun checkEmulatorProps(): Boolean {
        val props = listOf(
            "init.svc.qemud",
            "init.svc.qemu-props",
            "qemu.hw.mainkeys",
            "qemu.sf.fake_camera",
            "qemu.sf.lcd_density"
        )
        return props.any { propertyExists(it) }
    }

    private fun propertyExists(propertyName: String): Boolean {
        return try {
            val process = Runtime.getRuntime().exec("getprop $propertyName")
            process.inputStream.bufferedReader().readText().isNotEmpty()
        } catch (e: Exception) {
            false
        }
    }

    fun checkIsEmulator(): Boolean {
        val isEmulator = checkEmulatorProps()
        if (isEmulator) {
            crashlytics.recordException(Exception("Device Is Emulator"))
        }
        return isEmulator
    }

    fun getCheckEmulatorSetting(): Boolean {
        return sharedPreferences.getBoolean("checkEmulator", true)
    }

    fun setCheckEmulatorSetting(isChecked: Boolean) {
        val editor = sharedPreferences.edit()
        editor.putBoolean("checkEmulator", isChecked)
        editor.apply()
    }

    fun checkIsUSBDebugEnable(): Boolean {
        val isUSBDebugEnable = Settings.Secure.getInt(context.contentResolver, Settings.Global.ADB_ENABLED, 0) == 1
        if (isUSBDebugEnable) {
            crashlytics.recordException(Exception("USB Debugging Is Enable"))
        }
        return isUSBDebugEnable
    }

    fun getCheckUSBDebugSetting(): Boolean {
        return sharedPreferences.getBoolean("checkUSBDebug", true)
    }

    fun setCheckUSBDebugSetting(isChecked: Boolean) {
        val editor = sharedPreferences.edit()
        editor.putBoolean("checkUSBDebug", isChecked)
        editor.apply()
    }

    fun checkIsAccessibilityEnable(): Boolean {
        val am = context.getSystemService(Context.ACCESSIBILITY_SERVICE) as AccessibilityManager
        val isAccessibilityEnabled = am.isEnabled
        if (isAccessibilityEnabled) {
            crashlytics.recordException(Exception("Accessibility Service Is Enable"))
        }
        return isAccessibilityEnabled
    }

    fun getCheckAccessibilitySetting(): Boolean {
        return sharedPreferences.getBoolean("checkAccessibility", true)
    }

    fun setCheckAccessibilitySetting(isChecked: Boolean) {
        val editor = sharedPreferences.edit()
        editor.putBoolean("checkAccessibility", isChecked)
        editor.apply()
    }

    suspend fun saveUser(): String {
        val currentUser = auth.currentUser

        if (currentUser != null && currentUser.email != null) {
            val editor = sharedPreferences.edit()
            val docRef = firestoreDB.collection("users").document(
                currentUser.email ?: ""
            )
            // check data in firestore
            docRef.get().addOnCompleteListener { task: Task<DocumentSnapshot> ->
                if (task.isSuccessful) {
                    val document = task.result
                    if (document.exists()) {
                        editor.putString("username", document.getString("username"))
                        editor.putString("hash", document.getString("hash"))
                        editor.apply()
                    } else {
                        CoroutineScope(Dispatchers.IO).launch {
                            val call = apiService.connectUser()
                            if (call.isSuccessful) {
                                call.body()?.let {
                                    // success create data user from api
                                    // save data user to firestore
                                    val user: MutableMap<String, Any> = HashMap()
                                    user["username"] = it.username
                                    user["hash"] = it.hash
                                    docRef.set(user)
                                        .addOnSuccessListener { _ ->
                                            editor.putString("username", it.username)
                                            editor.putString("hash", it.hash)
                                            editor.apply()
                                        }
                                        .addOnFailureListener {
                                            throw Exception("Fail Sign In")
                                        }
                                }
                            } else {
                                throw Exception("Fail Sign In")
                            }
                        }
                    }
                } else {
                    throw Exception("Fail Sign In")
                }
            }
        } else {
            throw Exception("Fail Sign In")
        }

        return "Success Sign In"
    }

    suspend fun deleteUser() {
        auth.signOut()
        val editor = sharedPreferences.edit()
        editor.remove("username")
        editor.remove("hash")
        editor.apply()
        recipeDao.deleteAll()
    }

    fun getUserEmail(): String {
        auth.currentUser?.let {
            return it.email ?: "Email"
        } ?: kotlin.run {
            return "Email"
        }
    }

    fun getUserName(): String {
        auth.currentUser?.let {
            return it.displayName ?: "Name"
        } ?: kotlin.run {
            return "Name"
        }
    }

    fun getUserImage(): String {
        auth.currentUser?.let {
            return it.photoUrl.toString()
        } ?: kotlin.run {
            return ""
        }
    }

    suspend fun getRecipes(query: String): List<Recipe> {
        val resource = mutableListOf<Recipe>()
        try {
            val call = apiService.getRecipes(25, "popularity", query)
            if (call.code() != 200 && call.body() == null) {
                throw Exception("Data Empty")
            } else {
                call.body()?.let {
                    val recipeList = it.results.filter { recipe -> recipe.analyzedInstructions.isNotEmpty() && recipe.extendedIngredients.isNotEmpty() }
                    resource.addAll(recipeList)
                }
            }
        } catch (e: Exception) {
            throw Exception("No Data")
        }

        return resource
    }

    suspend fun addMealPlan(date: Long, slot: Int, recipe: Recipe): String {
        val result: String
        val username = sharedPreferences.getString("username", "") ?: ""
        val hash = sharedPreferences.getString("hash", "") ?: ""
        val valueBody =
            MealPlanValue(recipe.id, recipe.servings, recipe.title, recipe.image)
        val body = MealPlan(0L, date / 1000 + 25200, slot, 0, "RECIPE", valueBody)
        val call = apiService.addMealPlan(username, hash, body)
        result = if (call.isSuccessful) {
            "Success Add Meal Plan"
        } else {
            "Fail Add Meal Plan"
        }

        return result
    }

    suspend fun getRecipeInformation(id: Long): RecipeInformation {
        var result = RecipeInformation(
            0,
            "",
            "",
            "",
            "",
            0,
            0,
            0,
            listOf(),
            listOf(),
            false
        )
        val call = apiService.getRecipeInformation(id)
        if (call.code() == 200) {
            call.body()?.let {
                val recipeInformation = it
                val favoriteCount = recipeDao.getFavoriteRecipeById(
                    recipeInformation.id
                )
                if (favoriteCount != 0) {
                    recipeInformation.isFavorite = true
                }
                result = recipeInformation
            }
        } else {
            throw Exception("Fail Get Recipe Information")
        }
        return result
    }

    suspend fun deleteMealPlan(id: Long): String {
        val result: String
        val username = sharedPreferences.getString("username", "") ?: ""
        val hash = sharedPreferences.getString("hash", "") ?: ""
        val call = apiService.deleteMealPlan(username, id, hash)
        result = if (call.isSuccessful) {
            "Success Delete Meal Plan"
        } else {
            "Fail Delete Meal Plan"
        }
        return result
    }

    suspend fun getMealPlans(date: Date): List<MealPlan> {
        val mealPlan = mutableListOf<MealPlan>()
        val username = sharedPreferences.getString("username", "") ?: ""
        val hash = sharedPreferences.getString("hash", "") ?: ""
        val dateFormat = "yyyy-MM-dd"
        val formattedDate = SimpleDateFormat(dateFormat, Locale.getDefault()).format(date)
        val call = apiService.getMealPlans(username, formattedDate, hash)
        if (call.code() == 200) {
            call.body()?.let {
                mealPlan.addAll(it.items)
            }
        } else {
            throw Exception("No meals planned for that day")
        }
        return mealPlan
    }

    suspend fun addFavoriteRecipe(recipeEntity: RecipeEntity) {
        recipeDao.insert(recipeEntity)
    }

    suspend fun removeFavoriteRecipe(recipeEntity: RecipeEntity) {
        recipeDao.delete(recipeEntity)
    }

    fun getFavoriteRecipes(): LiveData<List<RecipeEntity>> = recipeDao.favoriteRecipes()
}
